[SPARK-36038][CORE] Speculation metrics summary at stage level

thejdeep opened the pull request:
### What changes were proposed in this pull request?
Currently there are no speculation metrics available for Spark either at application/job/stage level. This PR is to add some basic speculation metrics for a stage when speculation execution is enabled.

This is similar to the existing stage level metrics tracking numTotal (total number of speculated tasks), numCompleted (total number of successful speculated tasks), numFailed (total number of failed speculated tasks), numKilled (total number of killed speculated tasks) etc.

With this new set of metrics, it helps further understanding speculative execution feature in the context of the application and also helps in further tuning the speculative execution config knobs.

 ### Why are the changes needed?
Additional metrics for speculative execution.

 ### Does this PR introduce _any_ user-facing change?
Yes, Stages Page in SHS UI will have an additional table for speculation metrics, if present.

 ### How was this patch tested?
Unit tests added and also tested on our internal platform.

Absence of speculation metrics : 
![Screen Shot 2021-11-15 at 10 12 23 AM](https://user-images.githubusercontent.com/1708757/141836847-b0768265-35eb-4963-b83a-8b785318e82d.png)

Presence of speculation metrics : 
![Screen Shot 2021-11-15 at 10 11 52 AM](https://user-images.githubusercontent.com/1708757/141836945-1078bfcb-1c07-44db-9f33-5dbd09ece6f2.png)

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented:
ok to test

SparkQA commented:
**[Test build #145260 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145260/testReport)** for PR 34607 at commit [`ea6c7be`](https://github.com/apache/spark/commit/ea6c7be75c1b08407686680b112025f6e995e87b).

mridulm commented:
+CC @sarutak, @AngersZhuuuu who reviewed the last patch.

AngersZhuuuu commented:
Can you provide a diff between origin MR? I think it will make it easier to know what changed.

sarutak commented on core/src/main/resources/org/apache/spark/ui/static/stagepage.js:
```diff
@@ -652,6 +652,38 @@ $(document).ready(function () {
+        $("#speculationSummary").hide()
+        var speculationSummaryInfo = responseBody.speculationSummary;
+        var speculationData;
+        if(speculationSummaryInfo) {
```
The only difference from 73747ecb970595d49c478b0eb65 is this `if` block right?

thejdeep replied:
@sarutak Yes, that is the only difference.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49731/

sarutak commented:
@thejdeep Could you consider the following things?

* Please enable GitHub Actions on you repository.
* Could you follow the template in the description? Also, I think this PR is a retake of 73747ecb970595d49c478b0eb65f rather than a revert of the revert.

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49731/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/49731/

SparkQA commented:
**[Test build #145260 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145260/testReport)** for PR 34607 at commit [`ea6c7be`](https://github.com/apache/spark/commit/ea6c7be75c1b08407686680b112025f6e995e87b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/145260/

SparkQA commented:
**[Test build #145278 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145278/testReport)** for PR 34607 at commit [`8076599`](https://github.com/apache/spark/commit/80765993b30395780a0bc996749433459c773de1).

thejdeep committed:
### What changes were proposed in this pull request?
Currently there are no speculation metrics available for Spark either at application/job/stage level. This PR is to add some basic speculation metrics for a stage when speculation execution is enabled.

This is similar to the existing stage level metrics tracking numTotal (total number of speculated tasks), numCompleted (total number of successful speculated tasks), numFailed (total number of failed speculated tasks), numKilled (total number of killed speculated tasks) etc.

With this new set of metrics, it helps further understanding speculative execution feature in the context of the application and also helps in further tuning the speculative execution config knobs.

 ### Why are the changes needed?
Additional metrics for speculative execution.

 ### Does this PR introduce _any_ user-facing change?
Yes, Stages Page in SHS UI will have an additional table for speculation metrics, if present.

 ### How was this patch tested?
Unit tests added and also tested on our internal platform.

SparkQA commented:
**[Test build #145279 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145279/testReport)** for PR 34607 at commit [`79a591c`](https://github.com/apache/spark/commit/79a591cf08447ffc292f66a3ef209205a1c994d5).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49749/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49748/

thejdeep commented:
@sarutak Thanks for reviewing, updated the description and enabled workflows.

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49748/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/49748/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49749/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/49749/

SparkQA commented:
**[Test build #145279 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145279/testReport)** for PR 34607 at commit [`79a591c`](https://github.com/apache/spark/commit/79a591cf08447ffc292f66a3ef209205a1c994d5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/145279/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/145278/

SparkQA commented:
**[Test build #145284 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145284/testReport)** for PR 34607 at commit [`bcd32bf`](https://github.com/apache/spark/commit/bcd32bf51a7e434c925f1f9d859fc1b3a2fa4618).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49754/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49754/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/49754/

SparkQA commented:
**[Test build #145284 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145284/testReport)** for PR 34607 at commit [`bcd32bf`](https://github.com/apache/spark/commit/bcd32bf51a7e434c925f1f9d859fc1b3a2fa4618).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/145284/

AmplabJenkins commented:
Can one of the admins verify this patch?

sarutak commented on core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:
```diff
@@ -600,6 +600,12 @@ private[spark] class AppStatusListener(
+        stage.speculationStageSummary.numActiveTasks += 1
+        stage.speculationStageSummary.numTasks += 1
+      }
+      maybeUpdate(stage.speculationStageSummary, now)
```
Should we update only if the task is speculative?

thejdeep replied:
Not necessary, update can happen even if the current task is not speculative and if the update period is configured to a different time. 

What we could do is change it to `update` and not `maybeUpdate` if the task is speculative but those may be too many updates. What do you think ?

sarutak replied:
How about:
```suggestion
      if (stage.speculationStageSummary.numTasks > 0) {
        maybeUpdate(stage.speculationStageSummary, now)
      }
```
Otherwise, the metrics `speculationSummary` is always present even though a stage has no speculative tasks.
```
$ curl http://localhost:4040/api/v1/applications/local-1637337648156/stages/0/
...
  "speculationSummary" : {
    "numTasks" : 0,
    "numActiveTasks" : 0,
    "numCompletedTasks" : 0,
    "numFailedTasks" : 0,
    "numKilledTasks" : 0
  },
...
```

thejdeep replied:
ok, changed it.

sarutak commented on core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:
```diff
@@ -747,6 +753,19 @@ private[spark] class AppStatusListener(
+      if (isLastTask && event.taskInfo.speculative) {
+        update(speculationStageSummary, now)
+      } else {
+        maybeUpdate(speculationStageSummary, now)
```
ditto.

thejdeep replied:
Same response as above

sarutak commented on core/src/main/resources/org/apache/spark/ui/static/stagepage.js:
```diff
@@ -652,6 +652,38 @@ $(document).ready(function () {
           });
 
+        // Prepare data for speculation metrics
+        $("#speculationSummaryTitle").hide()
```
nit: Some statements added by this change doesn't end with `;`.

thejdeep replied:
updated it

venkata91 committed:
Author:    Venkata krishnan Sowrirajan <vsowrirajan@linkedin.com>

Commit:
Author:    Ron Hu <rhu@linkedin.com>

SparkQA commented:
**[Test build #145392 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145392/testReport)** for PR 34607 at commit [`950a55f`](https://github.com/apache/spark/commit/950a55f4af0da62aeed0cdcfa0be3750359dc367).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49865/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49865/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/49865/

SparkQA commented:
**[Test build #145392 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145392/testReport)** for PR 34607 at commit [`950a55f`](https://github.com/apache/spark/commit/950a55f4af0da62aeed0cdcfa0be3750359dc367).
 * This patch **fails from timeout after a configured wait of `500m`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/145392/

thejdeep commented:
@sarutak Can you please review again ? Thanks

thejdeep committed:
Only update speculation summary if stage has speculative tasks

SparkQA commented:
**[Test build #145459 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145459/testReport)** for PR 34607 at commit [`a883c83`](https://github.com/apache/spark/commit/a883c83d2d9688333d420f9a753c2d13ba7f1afc).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49931/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49931/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/49931/

SparkQA commented:
**[Test build #145459 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145459/testReport)** for PR 34607 at commit [`a883c83`](https://github.com/apache/spark/commit/a883c83d2d9688333d420f9a753c2d13ba7f1afc).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/145459/

sarutak commented on core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:
```diff
@@ -747,6 +755,19 @@ private[spark] class AppStatusListener(
+        speculationStageSummary.numFailedTasks += failedDelta
+        speculationStageSummary.numKilledTasks += killedDelta
+      }
+      if (isLastTask && event.taskInfo.speculative) {
+        update(speculationStageSummary, now)
+      } else if (stage.speculationStageSummary.numTasks > 0) {
+        maybeUpdate(speculationStageSummary, now)
+      }
```
```suggestion
      if (speculationStageSummary.numTasks > 0) {
        if (isLastTask) {
          update(speculationStageSummary, now)
        } else {
          maybeUpdate(speculationStageSummary, now)
        }
      }
```

thejdeep replied:
I changed `maybeUpdate` to `update` when we encounter a speculative task. The problem with the previous approach was that - if we have a speculative task, then future task end events would all make unnecessary speculation summary writes to DB. What do you think ?

thejdeep committed:
Change maybeUpdate to deterministic update when speculative task is encountered

SparkQA commented:
**[Test build #145470 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145470/testReport)** for PR 34607 at commit [`175d4e6`](https://github.com/apache/spark/commit/175d4e62d6180f416ab40821282ceb8064fef0ee).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49942/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49942/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/49942/

SparkQA commented:
**[Test build #145470 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145470/testReport)** for PR 34607 at commit [`175d4e6`](https://github.com/apache/spark/commit/175d4e62d6180f416ab40821282ceb8064fef0ee).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/145470/

thejdeep committed:
Fix speculation tests

SparkQA commented:
**[Test build #145472 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145472/testReport)** for PR 34607 at commit [`2ba73d0`](https://github.com/apache/spark/commit/2ba73d03b07eca48f5371aa93f18d5ac5a09225d).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49944/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49944/

SparkQA commented:
**[Test build #145472 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145472/testReport)** for PR 34607 at commit [`2ba73d0`](https://github.com/apache/spark/commit/2ba73d03b07eca48f5371aa93f18d5ac5a09225d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/145472/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/49944/

sarutak commented on core/src/main/scala/org/apache/spark/status/storeTypes.scala:
```diff
@@ -399,6 +399,20 @@ private[spark] class ExecutorStageSummaryWrapper(
+  @JsonIgnore @KVIndex("stage")
+  private def stage: Array[Int] = Array(stageId, stageAttemptId)
+
+  private[this] val id: Array[Int] = _id
```
This is not necessary is it?

thejdeep replied:
KVStore API requires an `id` but this was implemented just for convenience over class members and API constructs. I can remove `_id`

sarutak reviewed (commented):
> I changed maybeUpdate to update when we encounter a speculative task. The problem with the previous approach was that - if we have a speculative task, then future task end events would all make unnecessary speculation summary writes to DB. What do you think ?

How about changing `speculationStageSummary` in `LiveStage` to
`var speculationStageSummary: Option[LiveSpeculationStageSummary]` and then, we can check whether there is any update about speculation summary.
If it's `Some`, we can update with `maybeUpdate`.

thejdeep commented:
> > I changed maybeUpdate to update when we encounter a speculative task. The problem with the previous approach was that - if we have a speculative task, then future task end events would all make unnecessary speculation summary writes to DB. What do you think ?
> 
> How about changing `speculationStageSummary` in `LiveStage` to `var speculationStageSummary: Option[LiveSpeculationStageSummary]` and then, we can check whether there is any update about speculation summary. If it's `Some`, we can update with `maybeUpdate`.

@sarutak Would not really prefer it since it would require changing a class val to a var. I am assuming this is the change you suggested : 
```
AppStatusListener.scala

      if (event.taskInfo.speculative) {
        stage.speculationStageSummary = Some(stage.speculationStageSummary.getOrElse(
            new LiveSpeculationStageSummary(event.stageId, event.stageAttemptId)))
        val speculationStageSummary = stage.speculationStageSummary.get
        speculationStageSummary.numActiveTasks -= 1
        speculationStageSummary.numCompletedTasks += completedDelta
        speculationStageSummary.numFailedTasks += failedDelta
        speculationStageSummary.numKilledTasks += killedDelta
      }

      if(stage.speculationStageSummary.isDefined) {
        maybeUpdate(stage.speculationStageSummary.get, now)
      }
```
and 
```
LiveEntity.scala

var speculationStageSummary: Option[LiveSpeculationStageSummary] = None
```
What do you think ?

sarutak commented:
> since it would require changing a class val to a var.

What's the problem?

SparkQA commented:
**[Test build #145775 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145775/testReport)** for PR 34607 at commit [`d211952`](https://github.com/apache/spark/commit/d211952fb363b8329ab426994f9a2cfbd5dd7744).

thejdeep commented:
@sarutak Thanks, please review.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50248/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50248/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50248/

SparkQA commented:
**[Test build #145775 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145775/testReport)** for PR 34607 at commit [`d211952`](https://github.com/apache/spark/commit/d211952fb363b8329ab426994f9a2cfbd5dd7744).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/145775/

SparkQA commented:
**[Test build #145777 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145777/testReport)** for PR 34607 at commit [`c65fcee`](https://github.com/apache/spark/commit/c65fceea879aeb10a8be725557368e1065b2f09b).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50250/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50250/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50250/

SparkQA commented:
**[Test build #145777 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145777/testReport)** for PR 34607 at commit [`c65fcee`](https://github.com/apache/spark/commit/c65fceea879aeb10a8be725557368e1065b2f09b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/145777/

SparkQA commented:
**[Test build #145821 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145821/testReport)** for PR 34607 at commit [`b2319b8`](https://github.com/apache/spark/commit/b2319b808f636306f7895a357801998af3ec3b72).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50296/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50296/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50296/

SparkQA commented:
**[Test build #145821 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145821/testReport)** for PR 34607 at commit [`b2319b8`](https://github.com/apache/spark/commit/b2319b808f636306f7895a357801998af3ec3b72).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/145821/

SparkQA commented:
**[Test build #145832 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145832/testReport)** for PR 34607 at commit [`e9164fe`](https://github.com/apache/spark/commit/e9164fef0bb4aea5d1e990d4508560007544d5e2).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50307/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50307/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50307/

SparkQA commented:
**[Test build #145832 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145832/testReport)** for PR 34607 at commit [`e9164fe`](https://github.com/apache/spark/commit/e9164fef0bb4aea5d1e990d4508560007544d5e2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

thejdeep commented:
@sarutak This is the passing build https://github.com/thejdeep/spark/actions/runs/1528346418

mridulm commented:
Is there anything pending on this @sarutak ? Thx

sarutak commented on core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:
```diff
@@ -600,6 +600,17 @@ private[spark] class AppStatusListener(
+        speculationSummary.numTasks += 1
+      }
+      if(stage.speculationStageSummary.isDefined) {
+        maybeUpdate(stage.speculationStageSummary.get, now)
```
With the latest change, `stage.speculationStageSummary` never become `None` once it's substituted `Some`. So, unnecessary updates may happen line you were afraid.
https://github.com/apache/spark/pull/34607#discussion_r753559069

Let's change `maybeUpdate` to return whether it actually updates or not.
Then, we can set `None` to `stage.speculationStageSummary` if the update actually happens.

thejdeep replied:
How about we make a deterministic update if the taskInfo is speculative and then we avoid `maybeUpdate` altogether ?

thejdeep replied:
By setting it to `None`, we will be losing on accumulated data of speculation summary

thejdeep replied:
```      if (event.taskInfo.speculative) {
        stage.speculationStageSummary = Some(stage.speculationStageSummary
          .getOrElse(new LiveSpeculationStageSummary(event.stageId, event.stageAttemptId)))
        val speculationSummary = stage.speculationStageSummary.get
        speculationSummary.numActiveTasks += 1
        speculationSummary.numTasks += 1
       update(stage.speculationSummary, now)
      }```
This means we only update when the task is speculative. It will also leads to a minimum number of updates

sarutak replied:
> By setting it to None, we will be losing on accumulated data of speculation summary

Setting `None` means, currently updated speculation info is flushed.
Why updates will be lost?

thejdeep replied:
Since we are indexing by stageId and stageAttemptId, wont the second call to update overwrite the first ?
For example,

```
speculationSummary.numActiveTasks +=1
update(speculationSummary, now)

speculationSummary = new LiveStageSpeculationSummary(id, attempt)
update(speculationSummary, now)
```
The second update will overwrite the first one, right ? And for the stage, the speculation summary will show 0 tasks as the default value

sarutak replied:
Ah, OK, `stage.speculationStageSummary` hold just a delta.
So how about:

* Revert he declaration of `stage.speculationStageSummary` (`val speculationStageSummary: LiveSpeculationStageSummary = ...`)
* Declare a variable flag like `needUpdateSpeculationSummary` to indicate whether we need to call `maybeUpdate` or not.
* Set `true` to the flag when the summary is updated.
* Change `maybeUpdate` to return whether it has actually updated or not.
* set `false` to the flag when the `maybeUpdate` returns `true`.

mridulm replied:
Why are we special casing speculative execution summary compared to other updates (to stage, executor, etc).
With the change below (`maybeUpdate` only when task start is for speculative task) we should be fine ?

sarutak replied:
> With the change below (maybeUpdate only when task start is for speculative task) we should be fine ?
> With the change below (maybeUpdate only when task start is for speculative task) we should be fine ?

 `maybeUpdate` doesn't always write to the KVStore. So, if writing to the KVStore happens only when a speculative task starts, and the speculation summary of a speculative task is not written by `maybeUpdate`, the next chance the speculation summary can be written is the time another speculative task starts

mridulm replied:
Wont this not be the same for any of the other kv store updates in this method as well ? (just that the frequency is slightly higher than for speculative tasks).
Having said that, if that is a concern, replacing `maybeUpdate` with `update` - when done within `event.taskInfo.speculative` should be fine - it is a fairly infrequent occurrence anyay !

sarutak replied:
@thejdeep Are you satisfied with the idea @mridulm suggests, that `maybeUpdate` happens only if a task is speculative  ? You concerned about the similar idea.
https://github.com/apache/spark/pull/34607#discussion_r750963340

Another idea that changing `maybeUpdate` to `update` would one option.
I think it's inefficient for the HistoryServer but I'm OK with this idea if a stage should not start so many speculative tasks in the real-world use cases.

thejdeep replied:
@sarutak Agreed, we can just perform an update when we find a speculative task.

sarutak commented on core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:
```diff
@@ -747,6 +758,24 @@ private[spark] class AppStatusListener(
+        if (isLastTask) {
+          update(stage.speculationStageSummary.get, now)
+        } else {
+          maybeUpdate(stage.speculationStageSummary.get, now)
```
ditto.

sarutak commented:
I left a comment. Otherwise LGTM.

mridulm commented on core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:
```diff
@@ -600,6 +600,17 @@ private[spark] class AppStatusListener(
 
     Option(liveStages.get((event.stageId, event.stageAttemptId))).foreach { stage =>
+      if (event.taskInfo.speculative) {
+        stage.speculationStageSummary = Some(stage.speculationStageSummary
+          .getOrElse(new LiveSpeculationStageSummary(event.stageId, event.stageAttemptId)))
+        val speculationSummary = stage.speculationStageSummary.get
```
if (
```suggestion
  if (stage.speculationStageSummary.isEmpty) {
    // initialize
  }
  val speculationSummary = stage.speculationStageSummary.get
```

mridulm commented on core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:
```diff
@@ -600,6 +600,17 @@ private[spark] class AppStatusListener(
+        val speculationSummary = stage.speculationStageSummary.get
+        speculationSummary.numActiveTasks += 1
+        speculationSummary.numTasks += 1
+      }
+      if(stage.speculationStageSummary.isDefined) {
+        maybeUpdate(stage.speculationStageSummary.get, now)
```
We need to do the update only when it is speculative task start - not in other cases.
```suggestion
        maybeUpdate(stage.speculationStageSummary.get, now)
```

mridulm commented on core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:
```diff
@@ -747,6 +758,24 @@ private[spark] class AppStatusListener(
       }
 
+      if (event.taskInfo.speculative) {
+        stage.speculationStageSummary = Some(stage.speculationStageSummary.getOrElse(
+            new LiveSpeculationStageSummary(event.stageId, event.stageAttemptId)))
```
```suggestion
    if (stage. speculationStageSummary.isEmpty) {
      // initialize
      // Task Start was dropped - so we are going to see inconsistent metadata.
    }
```

mridulm commented on core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:
```diff
@@ -747,6 +758,24 @@ private[spark] class AppStatusListener(
+      if(stage.speculationStageSummary.isDefined) {
+        if (isLastTask) {
+          update(stage.speculationStageSummary.get, now)
+        } else {
```
```suggestion
        } else if (event.taskInfo.speculative) {
```

mridulm commented on core/src/main/scala/org/apache/spark/status/storeTypes.scala:
```diff
@@ -399,6 +399,18 @@ private[spark] class ExecutorStageSummaryWrapper(
+    val stageAttemptId: Int,
+    val info: SpeculationStageSummary) {
+
+  @JsonIgnore @KVIndex("stage")
+  private def stage: Array[Int] = Array(stageId, stageAttemptId)
+
```
We can drop this, and only keep `id`

mridulm commented on core/src/test/scala/org/apache/spark/status/AppStatusStoreSuite.scala:
```diff
@@ -141,6 +144,43 @@ class AppStatusStoreSuite extends SparkFunSuite {
 
+  test("SPARK-36038: speculation summary") {
+    val store = new InMemoryStore()
+    store.write(newSpeculationSummaryData(stageId, attemptId))
```
Keep the result of `newSpeculationStageSummary` in a variable and check against its fields in `info.foreach` below instead of hardcoding expected values.

mridulm reviewed (commented):
Took a quick pass.
@sarutak I have proposed some changes, to keep speculative summary updates inline with updates to executor/stage/etc already done in task start/end - please take a look !

@thejdeep I will let @sarutak drive the review given he has more context - was just reviewing on the fly :-)

thejdeep committed:
Address comments

thejdeep commented:
@sarutak Please review

SparkQA commented:
**[Test build #146076 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146076/testReport)** for PR 34607 at commit [`c0a8bf1`](https://github.com/apache/spark/commit/c0a8bf194f884edf29da1509e007cd9c7e46562e).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50550/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50550/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50550/

SparkQA commented:
**[Test build #146076 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146076/testReport)** for PR 34607 at commit [`c0a8bf1`](https://github.com/apache/spark/commit/c0a8bf194f884edf29da1509e007cd9c7e46562e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146076/

sarutak commented:
Merged to `master`. Thank you @thejdeep !
